// philips-hue/lib/philips-hue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _events = require("events");

var _events2 = _interopRequireDefault(_events);

var _fs = require("fs");

var _fs2 = _interopRequireDefault(_fs);

var _crypto = require("crypto");

var _crypto2 = _interopRequireDefault(_crypto);

var _axios = require("axios");

var _axios2 = _interopRequireDefault(_axios);

var _util = require("./util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var debug = require("debug")("philips-hue");

var PhilipsHue = function (_events$EventEmitter) {
  _inherits(PhilipsHue, _events$EventEmitter);

  function PhilipsHue() {
    _classCallCheck(this, PhilipsHue);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(PhilipsHue).call(this));

    _this.devicetype = 'node-philips-hue';
    _this.bridge = null;
    _this.username = null;
    return _this;
  }

  _createClass(PhilipsHue, [{
    key: "login",
    value: function login(confFile) {
      var _this2 = this;

      if (typeof confFile !== "string") return Promise.reject("Argument Error: config file is missing");
      debug("login file: " + confFile);
      try {
        if (_fs2.default.statSync(confFile).isFile()) {
          var conf = require(confFile);
          this.bridge = conf.bridge;
          this.username = conf.username;
          this.devicetype = conf.devicetype;
          return Promise.resolve(conf);
        }
      } catch (err) {
        debug("config file not exists");
      }
      debug("generate config file " + confFile);
      return this.getBridges().then(function (bridges) {
        debug("found bridges: " + JSON.stringify(bridges));
        _this2.bridge = bridges[0];
        if (!/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(_this2.bridge)) {
          throw "invalid bridge address \"" + _this2.bridge + "\"";
        }
        return _this2.auth(_this2.bridge);
      }).then(function (username) {
        _this2.username = username;
        var conf = { bridge: _this2.bridge, username: username, devicetype: _this2.devicetype };
        _fs2.default.writeFileSync(confFile, JSON.stringify(conf));
        return conf;
      });
    }
  }, {
    key: "getBridges",
    value: function getBridges() {
      debug("getBridges");
      return _axios2.default.get("https://www.meethue.com/api/nupnp").then(function (res) {
        (0, _util.checkResponse)(res.data);
        return res.data.map(function (i) {
          return i.internalipaddress;
        });
      });
    }
  }, {
    key: "auth",
    value: function auth(bridge) {
      debug("auth bridge: " + bridge);
      return (0, _axios2.default)({
        method: "post",
        url: "http://" + bridge + "/api",
        data: JSON.stringify({
          devicetype: this.devicetype
        })
      }).then(function (res) {
        debug(res.data);
        (0, _util.checkResponse)(res.data);
        return res.data[0].success.username;
      });
    }

    // Bridge API request

  }, {
    key: "request",
    value: function request(opts) {
      var url = "http://" + this.bridge + "/api/" + this.username;
      return (0, _axios2.default)({
        url: "" + url + opts.path,
        method: opts.method || 'get',
        data: opts.data ? JSON.stringify(opts.data) : null
      }).then(function (res) {
        debug(res.data);
        (0, _util.checkResponse)(res.data);
        return res.data;
      });
    }
  }]);

  return PhilipsHue;
}(_events2.default.EventEmitter);

exports.default = PhilipsHue;
